File: src/test/java/some/Note.java
package some;

import io.github.petrknap.persistence.zoneddatetime.UtcWithLocal;
import io.github.petrknap.persistence.zoneddatetime.UtcWithTimezone;
import io.github.petrknap.persistence.zoneddatetime.UtcDateTimeConverter;
import jakarta.persistence.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.ZoneId;
import java.time.ZonedDateTime;

@Entity
@Table(name = "notes")
final public class Note
{
    @Id
    @Column
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private @Nullable Long id;

    /**
     * Example: UTC date-time with local date-time
     */
    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "utc", column = @Column(name = "created_at__utc")),
            @AttributeOverride(name = "local", column = @Column(name = "created_at__local"))
    })
    private @NotNull UtcWithLocal createdAt;

    /**
     * Example: UTC date-time with timezone
     */
    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "utc", column = @Column(name = "created_at_2__utc")),
            @AttributeOverride(name = "timezone", column = @Column(name = "created_at_2__timezone"))
    })
    private @NotNull UtcWithTimezone createdAt2;

    /**
     * Example: nullable embeddable
     */
    @Embedded
    private @Nullable UtcWithLocal deletedAt;

    /**
     * Example: UTC date-time converter
     */
    @Column(nullable = false)
    @Convert(converter = UtcDateTimeConverter.class)
    public @NotNull ZonedDateTime createdAtUtc;

    /**
     * Example: converted nullable
     */
    @Column
    @Convert(converter = UtcDateTimeConverter.class)
    public @Nullable ZonedDateTime deletedAtUtc;

    @Column(nullable = false)
    private @NotNull String content;

    public Note(
            @NotNull ZonedDateTime createdAt,
            @NotNull String content
    ) {
        this.createdAt = new UtcWithLocal(createdAt);
        this.createdAt2 = new UtcWithTimezone(createdAt);
        this.createdAtUtc = createdAt.withZoneSameInstant(ZoneId.of("UTC"));
        this.content = content;
    }

    private Note() {}

    public @NotNull ZonedDateTime getCreatedAt() {
        return createdAt.toZonedDateTime();
    }

    public @NotNull ZonedDateTime getCreatedAt2() {
        return createdAt2.toZonedDateTime();
    }

    public @Nullable ZonedDateTime getDeletedAt() {
        return deletedAt != null ? deletedAt.toZonedDateTime() : null;
    }
}

File: src/test/php/Some/Note.php
<?php

declare(strict_types=1);

namespace PetrKnap\ZonedDateTimePersistence\Some;

use DateTimeImmutable;
use DateTimeInterface;
use DateTimeZone;
use Doctrine\ORM\Mapping as ORM;
use PetrKnap\ZonedDateTimePersistence\UtcWithLocal;
use PetrKnap\ZonedDateTimePersistence\UtcWithTimezone;
use PetrKnap\ZonedDateTimePersistence\UtcDateTimeType;

#[ORM\Entity]
#[ORM\HasLifecycleCallbacks]
#[ORM\Table(name: 'notes')]
final class Note
{
    #[ORM\Id]
    #[ORM\Column]
    #[ORM\GeneratedValue(strategy: 'IDENTITY')]
    protected int|null $id = null;

    /**
     * Example: utc date-time with local date-time
     */
    #[ORM\Embedded(columnPrefix: 'created_at__')]
    protected UtcWithLocal $createdAt;

    /**
     * Example: utc date-time with timezone identifier
     */
    #[ORM\Embedded(columnPrefix: 'created_at_2__')]
    protected UtcWithTimezone $createdAt2;

    /**
     * Example: nullable embeddable
     */
    #[ORM\Embedded(columnPrefix: 'deleted_at__')]
    protected UtcWithLocal|null $deletedAt = null;

    /**
     * Example: utc date-time
     */
    #[ORM\Column(name: 'created_at_utc', type: UtcDateTimeType::NAME, options: ['default' => 'CURRENT_TIMESTAMP'])]
    public DateTimeInterface $createdAtUtc;

    /**
     * Example: nullable type
     */
    #[ORM\Column(name: 'deleted_at_utc', type: UtcDateTimeType::NAME, nullable: true)]
    public DateTimeInterface|null $deletedAtUtc = null;

    public function __construct(
        DateTimeImmutable $createdAt,
        #[ORM\Column(name: 'content', nullable: false)]
        protected string $content,
    ) {
        $this->createdAt = new UtcWithLocal($createdAt);
        $this->createdAt2 = new UtcWithTimezone($createdAt);
        $this->createdAtUtc = $createdAt->setTimezone(new DateTimeZone('UTC'));
    }

    /**
     * @internal this is an event listener
     */
    #[ORM\PostLoad]
    public function fixNullables(): void
    {
        $this->deletedAt = $this->deletedAt?->asNullable();
    }

    public function getContent(): string
    {
        return $this->content;
    }

    public function getCreatedAt(): DateTimeInterface
    {
        return $this->createdAt->toZonedDateTime();
    }

    public function getDeletedAt(): DateTimeInterface|null
    {
        return $this->deletedAt?->toZonedDateTime();
    }
}
